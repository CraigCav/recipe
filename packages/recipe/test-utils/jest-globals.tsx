import React from 'react';
import {render, cleanup} from '@testing-library/react';
import {ThemeProvider} from 'emotion-theming';
import {standard} from '../themes';

import '@testing-library/jest-dom/extend-expect';

const renderWithTheme = renderFn => (component, ...rest) =>
  renderFn(<ThemeProvider theme={standard}>{component}</ThemeProvider>, rest);

export const renderToHtml = renderWithTheme((ui, options) => {
  const {container} = render(ui, options);
  return container.outerHTML;
});

export const fullRender = renderWithTheme(render);

afterEach(cleanup);
