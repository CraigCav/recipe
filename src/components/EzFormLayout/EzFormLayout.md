---
name: EzFormLayout
title: Form layout
category: forms
path: '/components/ez-form-layout'
---

Form layouts are used to arrange fields within a form using standard spacing. Fields are stacked vertically, but can be further grouped into rows using the `EzLayout` component.

---

## Best Practices

Forms should:

- Follow a predicable order, guiding the user to fill out the most important content first
- Group related tasks under section titles to provide more context and make the form easier to scan

---

## Examples

### Standard form layout

```jsx
<EzFormLayout>
  <EzField type="input" label="First name" />
  <EzField type="input" label="Last name" />
  <EzField type="number" label="Favorite Number" value={7} />
</EzFormLayout>
```

### Field groups

Use `EzLayout` to arrange multiple fields into a row or to vary form field layout across various breakpoints.

Typically used to group familiar layouts such as state and zip code, or start/end dates. Fields that are not directly related to each other should not be grouped next to each other as this makes it easier to miss at a glance.

```jsx
<EzFormLayout>
  <EzLayout layout="basic">
    <EzField type="input" label="First name" />
    <EzField type="input" label="Last name" />
  </EzLayout>
</EzFormLayout>
```

---

## Related components

- [EzField](/components/ez-field)
- [EzLayout](/components/ez-layout)
- [EzAlert](/components/ez-alert)
