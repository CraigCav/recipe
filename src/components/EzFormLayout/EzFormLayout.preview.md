```jsx
<EzFormLayout>
  <EzField type="text" label="■■■■■■■■■" />
  <EzField type="text" label="■■■■■■■■■" />
  <EzField type="number" label="■■■■■■■■■" value="■■" />
</EzFormLayout>
```