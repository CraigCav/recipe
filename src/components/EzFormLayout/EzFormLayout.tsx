import styled from '../../themes/styled';

const EzFormLayout = styled.div`
  > *:not(:first-child) {
    margin-top: ${({theme}) => theme.spacing.xl};
  }
`;

export default EzFormLayout;
