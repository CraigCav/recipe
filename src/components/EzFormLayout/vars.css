:root {
  --recipe-form-layout-field-margin: var(--recipe-global-static-size-300);
}