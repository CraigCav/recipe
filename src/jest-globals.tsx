import {shallow as enzymeShallow, render as enzymeRender, mount as enzymeMount} from 'enzyme';
import React from 'react';
import {create as reactCreate} from 'react-test-renderer';
import {render as rtlRender, cleanup} from 'react-testing-library';
import {renderToStaticMarkup} from 'react-dom/server';
import {ThemeProvider} from 'emotion-theming';
import {standard} from './themes';

import '@testing-library/jest-dom/extend-expect';

const renderWithTheme = renderFn => (component, ...rest) =>
  renderFn(<ThemeProvider theme={standard}>{component}</ThemeProvider>, rest);

export const shallow = tree => {
  const wrapper = enzymeShallow(<ThemeProvider theme={standard} />);
  const instance = wrapper.instance() as any;
  const context = instance.getChildContext();

  return enzymeShallow(tree, {context});
};

export const mount = tree => {
  const wrapper = enzymeShallow(<ThemeProvider theme={standard} />);
  const instance = wrapper.instance() as any;
  const context = instance.getChildContext();

  return enzymeMount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};

export const renderToHtml = renderWithTheme(renderToStaticMarkup);

export const render = renderWithTheme(enzymeRender);

export const fullRender = renderWithTheme(rtlRender);

export const create = renderWithTheme(reactCreate);

afterEach(cleanup);
