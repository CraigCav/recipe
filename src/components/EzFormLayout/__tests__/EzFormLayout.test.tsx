import React from 'react';
import EzFormLayout from '../EzFormLayout';
import {EzField} from '../../';
import {axe} from 'jest-axe';

import {fullRender, renderToHtml} from '../../../jest-globals';

describe('EzFormLayout', () => {
  const SampleLayout = () => (
    <EzFormLayout>
      <EzField label="First field" />
      <EzField label="Second field" />
    </EzFormLayout>
  );

  it('should render with default styles', () => {
    const actual = fullRender(<SampleLayout />);
    expect(actual.baseElement).toMatchSnapshot();
  });

  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(<SampleLayout />);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
