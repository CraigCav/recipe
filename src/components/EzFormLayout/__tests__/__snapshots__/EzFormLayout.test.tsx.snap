// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EzFormLayout should render with default styles 1`] = `
.recipe-8 > *:not(:first-child) {
  margin-top: 1.7142857142857142rem;
}

.recipe-2 {
  position: relative;
}

.recipe-2 > input,
.recipe-2 > textarea {
  display: block;
  width: 100%;
  background: #ffffff;
  border: 1px solid #ced4d9;
  border-radius: 4px;
  padding: 0.5714285714285714rem 0.8571428571428571rem;
  margin-top: 0.5714285714285714rem;
}

.recipe-0 {
  color: #565a5c;
  display: block;
  font-size: 1rem;
  font-weight: 700;
}

<body>
  <div>
    <div
      class="recipe-8 recipe-9"
    >
      <div
        class="recipe-2 recipe-3"
      >
        <label
          class="recipe-0 recipe-1"
          for="control$1"
        >
          First field
        </label>
        <input
          id="control$1"
          label="First field"
          name="control$1"
          type="text"
        />
      </div>
      <div
        class="recipe-2 recipe-3"
      >
        <label
          class="recipe-0 recipe-1"
          for="control$2"
        >
          Second field
        </label>
        <input
          id="control$2"
          label="Second field"
          name="control$2"
          type="text"
        />
      </div>
    </div>
  </div>
</body>
`;
